### PR TITLE
shared-informer: send sync updates to all handlers if the store was updated

### PR DESCRIPTION
**Pull request by squeed**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
informers: Don't treat relist same as sync

This is an alternative to the solution proposed in #86015 

It was observed that notifications can be missed when an object is changed during connection disruption if the handler is not set to receive resyncs. This case can be hit when either a handler has no ResyncPeriod, or when it's next resync deadline is after that of another handler on the same informer.

The solution is to compare the new object with the store in all cases, and treat it as a real Update if it has changed.

**Does this PR introduce a user-facing change?**:
```release-note
Shared informers are now more reliable in the face of network disruption.
```

/cc @sttts
/cc @liggitt

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 1]
```diff
@@ -453,9 +454,13 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
...
 		switch d.Type {
 		case Sync, Added, Updated:
-			isSync := d.Type == Sync
 			s.cacheMutationDetector.AddObject(d.Object)
 			if old, exists, err := s.indexer.Get(d.Object); err == nil && exists {
+				// The processor wants to know if this is a resync or a real
+				// object change; it's a resync if the Delta says so and our
+				// store is up-to-date. Sync is also emitted by DeltaFIFO on
+				// Relist, so it's not guaranteed to be a no-op change.
+				isSync := d.Type == Sync && isObjectSame(old, d.Object)
```
Comment (sttts) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: isResync

File: staging/src/k8s.io/client-go/tools/cache/shared_informer.go [hunk 2]
```diff
@@ -699,3 +704,28 @@ func (p *processorListener) setResyncPeriod(resyncPeriod time.Duration) {
...
+	if !ok {
+		return false
+	}
+
+	newRO, ok := newObj.(runtime.Object)
+	if !ok {
+		return false
+	}
+
+	oldRV, err := meta.NewAccessor().ResourceVersion(oldRO)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/tools/cache/shared_informer.go]: can we construct a single accessor outside the function as a private package var and reuse it?

**Comment by squeed** [comment]
Thanks for the review, updated.

**Comment by liggitt** [comment]
My preference is not to compensate downstream of DeltaFIFO like this, and @deads2k agrees.

I think we can find a way to tweak the first PR to fix the issue without breaking other DeltaFIFO clients. Let's sync up there (pun fully intended).

/close

**Comment by squeed** [comment]
Hmm, I actually liked solving it here better, since only the SharedInformer really understands the concept between handlers that do and do not want resyncs.

**Comment by liggitt** [comment]
> only the SharedInformer really understands the concept of handlers that do and do not want resyncs

sure, but making the shared informer reverse engineer whether this was a resync or a relist is not necessary if we make the deltafifo just tell us which it was

**Pull request closed without merging**